src/klein/resource.py

```diff
@@ -75,7 +75,10 @@ def _extractURLparts(request):
     @rtype: L{tuple} of L{unicode}, L{unicode}, L{int}, L{unicode}, L{unicode}
     """
     server_name = request.getRequestHostname()
-    server_port = request.getHost().port
+    if hasattr(request.getHost(), 'port'):
+        server_port = request.getHost().port
+    else:
+        server_port = 0
```

Comment:
I think the correct behaviour here is to leave `server_name` unchanged if there's no port rather than treating it as port 0.